### PR TITLE
chore: use custom tokenizer

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR uses a custom, production-ready tokenizer for CSS selectors based on [parsel](https://parsel.verou.me/).

### CODE REVIEW

File: packages/puppeteer-core/src/injected/PSelectorTokenizer.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
+import {assert} from '../util/assert.js';
```
Comment (OrKoN) [inline packages/puppeteer-core/src/injected/PSelectorTokenizer.ts:23]: Provide appropriate license

File: packages/puppeteer-core/src/injected/PSelectorTokenizer.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+          case TokenType.Comma:
+            token.content = token.content.trim() || ' ';
+            break;
+        }
+        break;
+    }
+  }
+}
+
+export function tokenize(selector: string, grammar = TOKENS): Token[] {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/injected/PSelectorTokenizer.ts:106]: Include tests for the tokenize functions with most common cases

**Review by OrKoN — CHANGES_REQUESTED** [review]

File: packages/puppeteer-core/src/injected/PSelectorTokenizer.ts (hunk 2 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/injected/PSelectorTokenizer.ts:106]: The tests for P selectors should sufficiently cover the tokenizer.

File: test/src/tokenizer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import expect from 'expect';
```
Comment (OrKoN) [inline test/src/tokenizer.spec.ts:17]: Missing the license in this file

File: test/src/tokenizer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import {tokenize} from 'puppeteer-core/internal/injected/PSelectorTokenizer.js';
+
+describe('PSelectorTokenizer', () => {
```
Comment (OrKoN) [inline test/src/tokenizer.spec.ts:20]: Let's add some parsing tests for child, sibling, general sibling, adjacent sibling and pseudo elements?

File: test/src/tokenizer.spec.ts (hunk 3 above)
Comment (jrandolf) [reply test/src/tokenizer.spec.ts:17]: No license needed. These tests are not from Parsel.

File: test/src/tokenizer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    expect(
+      JSON.stringify(tokenize('.mb-\\[max\\(-70\\%\\2c -23rem\\)\\]'))
+    ).toStrictEqual(
+      '[{"name":"mb-\\\\[max\\\\(-70\\\\%\\\\2c","type":"class","content":".mb-\\\\[max\\\\(-70\\\\%\\\\2c","pos":[0,19]},{"type":"combinator","content":" ","pos":[19,20]},{"name":"-23rem\\\\)\\\\]","type":"type","content":"-23rem\\\\)\\\\]","pos":[20,30]}]'
+    );
+  });
+
+  it('should work with complex selectors', () => {
+    expect(
+      JSON.stringify(tokenize('a > b, c ~ d, e ::before ::after(a)'))
```
Comment (OrKoN) [inline test/src/tokenizer.spec.ts]: nit, `a + b`

File: test/src/tokenizer.spec.ts (hunk 3 above)
Comment (OrKoN) [reply test/src/tokenizer.spec.ts:17]: Google-license is needed.

File: test/src/tokenizer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        tokenize(
+          '[data-test-id^="test-"]:not([data-test-id^="test-foo"]) [data-test-id^="test-"]:not([data-test-id^="test-foo"])'
+        )
+      )
+    ).toStrictEqual(
+      '[{"name":"data-test-id","operator":"^=","value":"\\"test-\\"","type":"attribute","content":"[data-test-id^=\\"test-\\"]","pos":[0,23]},{"name":"not","argument":"[data-test-id^=\\"test-foo\\"]","type":"pseudo-class","content":":not([data-test-id^=\\"test-foo\\"])","pos":[23,55]},{"type":"combinator","content":" ","pos":[55,56]},{"name":"data-test-id","operator":"^=","value":"\\"test-\\"","type":"attribute","content":"[data-test-id^=\\"test-\\"]","pos":[56,79]},{"name":"not","argument":"[data-test-id^=\\"test-foo\\"]","type":"pseudo-class","content":":not([data-test-id^=\\"test-foo\\"])","pos":[79,111]}]'
+    );
+  });
+
+  it('should work with CSS escapes', () => {
```
Comment (OrKoN) [inline test/src/tokenizer.spec.ts:53]: let's add a test for invalid selectors too? E.g., missing bracket balance etc

File: test/src/tokenizer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import expect from 'expect';
+import {tokenize} from 'puppeteer-core/internal/injected/PSelectorTokenizer.js';
+
+describe('PSelectorTokenizer', () => {
+  it('should work', () => {
```
Comment (OrKoN) [inline test/src/tokenizer.spec.ts:21]: does it work with `*`?

**Review by OrKoN — APPROVED** [review]

**Comment by OrKoN** [comment]
Flakiness looks unrelated

**Pull request merged**